json.go

```diff
@@ -0,0 +1,186 @@
+package matrixfederation
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"sort"
+	"unicode/utf8"
+)
+
+// CanonicalJSON re-encodes the JSON in a cannonical encoding. The encoding is
+// the shortest possible encoding using integer values with sorted object keys.
+func CanonicalJSON(input []byte) ([]byte, error) {
```

Comment:
Do we have a spec for this? Can you link to it in the comments please? E.g. http://matrix.org/docs/spec/server_server/unstable.html#canonical-json ?